src/core_init_atmosphere/mpas_kd_tree.F

```diff
@@ -0,0 +1,403 @@
+         distance = current_distance
+         res => kdtree
+      endif
+
+      !
+      ! To find the nearest neighbor, first serach the tree in a similar manner
+      ! as a single dimensioned BST, by comparing points on the current split
+      ! dimension.
+      !
+      ! If the distance between the current node and the query is less then the
+      ! minimum distance found within the subtree we just searched, then the nearest
+      ! neighbor might be in the opposite subtree, so search it.
+      !
+
+      if (query(kdtree % split_dim) > kdtree % point(kdtree % split_dim)) then
+         if (associated(kdtree % right)) then ! Search right
+            call mpas_kd_search_internal(kdtree % right, query, res, distance)
+         endif
+         if ((kdtree % point(kdtree % split_dim) - query(kdtree % split_dim))**2 <= distance .and. associated(kdtree % left)) then
+            call mpas_kd_search_internal(kdtree % left, query, res, distance) ! Check the other subtree
+         endif
+      else if (query(kdtree % split_dim) < kdtree % point(kdtree % split_dim)) then
+         if (associated(kdtree % left)) then ! Search left
+            call mpas_kd_search_internal(kdtree % left, query, res, distance)
+         endif
+         if ((kdtree % point(kdtree % split_dim) - query(kdtree % split_dim))**2 <= distance .and. associated(kdtree % right)) then
+            call mpas_kd_search_internal(kdtree % right, query, res, distance) ! Check the other subtree
+         endif
+      else ! Nearest point could be in either left or right subtree, so search both
+         if (associated(kdtree % right)) then
+            call mpas_kd_search_internal(kdtree % right, query, res, distance)
+         endif
+         if (associated(kdtree % left)) then
+            call mpas_kd_search_internal(kdtree % left, query, res, distance)
+         endif
+      endif
+
+   end subroutine mpas_kd_search_internal
+
+   !***********************************************************************
+   !
+   !  routine mpas_kd_search
+   !
+   !> \brief   Find the nearest point in a KD-Tree to a query
+   !> \author  Miles A. Curry
+   !> \date    01/28/20
+   !> \details
+   !> Search kdtree and returned the nearest point to query into the
+   !> res argument. Optionally, if distance is present, returned the
+   !> squared distance between query and res.
+   !>
+   !> If the dimension of query does not match the dimensions of points
+   !> within kdtree, then res will be returned as unassociated.
+   !
+   !-----------------------------------------------------------------------
+   subroutine mpas_kd_search(kdtree, query, res, distance)
+
+      implicit none
+      type (mpas_kd_type), pointer, intent(in) :: kdtree
+      real (kind=RKIND), dimension(:), intent(in) :: query
+      type (mpas_kd_type), pointer, intent(inout) :: res
+      real (kind=RKIND), intent(out), optional :: distance
+
+      real (kind=RKIND) :: dis
+
```

Comment:
If the kd-tree was constructed with zero points, then the tree pointer returned by `mpas_kd_construct` will be an unassociated pointer. If this is then passed to `mpas_kd_search`, a segfault will likely result. We could add a check like
```fortran
     if (.not. associated(kdtree)) then
        res => null()
        return
     end if
```
here in the `mpas_kd_search` routine to protect against this. We could then add a note in the description for `mpas_kd_search` that if no nearest neighbor was found, an unassociated `res` pointer is returned (with the only likely circumstance under which this would occur being an empty tree).

Reply:
That sounds like a great idea, I'll go ahead and implement that.